File: test/leetcode/numberComplement/number_complement_test.go
package numberComplement_test

import (
	"github.com/Ydot19/epi/leetcode/numberComplement"
	"github.com/stretchr/testify/suite"
	"testing"
)

func TestNumberComplement(t *testing.T) {
	suite.Run(t, new(NumberComplementTestSuite))
}

type NumberComplementTestSuite struct {
	suite.Suite
	Impl numberComplement.NumberComplement
}

func (suite *NumberComplementTestSuite) SetupTest() {
	suite.Impl = numberComplement.NewNumberComplement()
}

func (suite *NumberComplementTestSuite) Cases(f func(num int) int) {
	type testCase struct {
		description string
		num         int
		expected    int
	}

	for _, scenario := range []testCase{
		{
			description: "Case 1",
			num:         5,
			expected:    2,
		},
		{
			description: "Case 2",
			num:         2147483646,
			expected:    1,
		},
		{
			description: "Case 3",
			num:         319,
			expected:    192,
		},
	} {
		suite.T().Run(scenario.description, func(t *testing.T) {
			actual := f(scenario.num)
			suite.Equal(scenario.expected, actual)
		})
	}
}

func (suite *NumberComplementTestSuite) TestSolutionA() {
	suite.Cases(suite.Impl.SolutionA)
}

func (suite *NumberComplementTestSuite) TestOptimized() {
	suite.Cases(suite.Impl.Optimized)
}

File: test/leetcode/numberComplement/number_complement_test.ts
import {assertEquals} from "https://deno.land/std@0.119.0/testing/asserts.ts";
import {NumberComplement} from "../../../leetcode/numberComplement/number_complement.ts"

const Impl: NumberComplement = new NumberComplement()

type testCase = {
    description: string
    num: number
    expected: number
}

const testCases: Array<testCase> = [
    {
        description:"Case 1",
        num:5,
        expected: 2,
    },
    {
        description: "Case 2",
        num:         2147483646,
        expected:    1,
    },
    {
        description: "Case 3",
        num:         319,
        expected:    192,
    }
]

Deno.test("Test Number Complement Implementation", async(t)=>{
    const testFuncs = [
        Impl.solutionA,
        Impl.solutionB,
        Impl.optimized
    ]

    for (const func of testFuncs) {
        for (const tc of testCases){
            const description = `${func.name}: ${tc.description}`
            await t.step(description, ()=>{
                // act
                const actual = func(tc.num)
                // assert
                assertEquals(tc.expected, actual)
            })
        }
    }
})
